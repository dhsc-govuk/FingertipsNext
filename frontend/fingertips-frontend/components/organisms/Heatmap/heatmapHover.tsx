import { GovukColours } from '@/lib/styleHelpers/colours';
import { Paragraph } from 'govuk-react';
import { JSX } from 'react';
import styled from 'styled-components';
import { typography } from '@govuk-react/lib';
import { HeatmapBenchmarkProps } from './heatmapUtil';
import { HeatmapHoverBenchmarkPill } from './heatmapHoverBenchmarkPill';

const StyledDivHover = styled.div({
  color: GovukColours.Black,
  backgroundColor: GovukColours.White,
  boxShadow: `0px 0px 4px 0px ${GovukColours.DarkGrey}`,
  borderRadius: '8px',
  display: 'none',
  zIndex: 1,
  position: 'fixed',
  textAlign: 'left',
  padding: '16px',
  whiteSpace: 'normal',
  overflow: 'visible',
  maxWidth: '240px',
  transform: 'translateY(-50%)',
});

const StyledDivTriangle = styled.div({
  backgroundColor: GovukColours.White,
  boxShadow: `0px 0px 4px 0px ${GovukColours.DarkGrey}`,
  borderRadius: '2px',
  display: 'block',
  zIndex: 1,
  position: 'absolute',
  width: '16px',
  height: '16px',
  left: '-8px',
  top: '50%',
  transform: 'translateX(1px) translateY(-8px) rotate(45deg)',
  border: 'none',
});

const StyledDivTriangleOccluder = styled.div({
  backgroundColor: GovukColours.White,
  display: 'block',
  zIndex: 2,
  position: 'absolute',
  width: '16px',
  height: '24px',
  left: '0px',
  top: '50%',
  transform: 'translateY(-12px)',
});

const StyledParagraph = styled(Paragraph)(typography.font({ size: 16 }));
const StyledParagraphZeroMargin = styled(StyledParagraph)({
  marginBottom: '0',
});

export interface HeatmapHoverProps {
  areaName: string;
  period: number;
  indicatorName: string;
  value?: number;
  unitLabel: string;
  benchmark: HeatmapBenchmarkProps;
  hoverId?: string;
}

export function HeatmapHover({
  areaName,
  period,
  indicatorName,
  value,
  unitLabel,
  benchmark,
  hoverId,
}: Readonly<HeatmapHoverProps>): JSX.Element {
  return (
    <StyledDivHover id={hoverId}>
      <StyledDivTriangle />
      <StyledDivTriangleOccluder />
      <StyledParagraphZeroMargin>{`**${areaName}**`}</StyledParagraphZeroMargin>
      {period ? (
        <StyledParagraphZeroMargin>{`${period}`}</StyledParagraphZeroMargin>
      ) : null}
      <StyledParagraph>{indicatorName}</StyledParagraph>
      <HeatmapHoverBenchmarkPill
        value={value}
        unitLabel={unitLabel}
        outcome={benchmark.outcome}
        benchmarkMethod={benchmark.benchmarkMethod}
        polarity={benchmark.polarity}
      />
    </StyledDivHover>
  );
}
