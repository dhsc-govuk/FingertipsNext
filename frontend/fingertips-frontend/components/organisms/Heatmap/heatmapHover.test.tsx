import { render } from '@testing-library/react';
import { HeatmapHover, HeatmapHoverProps } from './heatmapHover';
import {
  BenchmarkComparisonMethod,
  BenchmarkOutcome,
  IndicatorPolarity,
} from '@/generated-sources/ft-api-client';

const testValue = 67;
const testHoverProps: HeatmapHoverProps = {
  areaName: 'Sesame Street',
  period: 1966,
  indicatorName:
    'Cookies eaten by the cookie monster compared to total US production',
  value: testValue,
  unitLabel: '%',
  benchmark: {
    outcome: BenchmarkOutcome.Better,
    benchmarkMethod: BenchmarkComparisonMethod.CIOverlappingReferenceValue99_8,
    polarity: IndicatorPolarity.HighIsGood,
  },
};

describe('heatmap hover', () => {
  it('should display area, period, and indicator name', () => {
    // TODO why is this rendering twice?
    const screen = render(
      <HeatmapHover
        areaName={testHoverProps.areaName}
        period={testHoverProps.period}
        indicatorName={testHoverProps.indicatorName}
        unitLabel={testHoverProps.unitLabel}
        benchmark={testHoverProps.benchmark}
      />
    );

    expect(screen.getByText(testHoverProps.areaName)).toBeInTheDocument();
    expect(screen.getByText(testHoverProps.period)).toBeInTheDocument();
    expect(screen.getByText(testHoverProps.indicatorName)).toBeInTheDocument();
  });
});

describe('snapshot', () => {
  const container = render(
    <HeatmapHover
      areaName={testHoverProps.areaName}
      period={testHoverProps.period}
      indicatorName={testHoverProps.indicatorName}
      unitLabel={testHoverProps.unitLabel}
      benchmark={testHoverProps.benchmark}
    />
  );
  expect(container.asFragment()).toMatchSnapshot();
});
