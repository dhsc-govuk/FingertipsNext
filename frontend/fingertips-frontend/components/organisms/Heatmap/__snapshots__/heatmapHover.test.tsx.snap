// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  .c3 {
  margin: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 15px;
}

.c3 > p {
  margin: 0;
}

.c3 > p > code {
  padding: 0.2em 0.4em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(27,31,35,0.05);
  border-radius: 3px;
}

.c3 > pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f6f8fa;
  border-radius: 3px;
}

.c3 > pre > code {
  display: inline;
  padding: 0;
  margin: 0;
  border: 0;
  overflow: visible;
  line-height: inherit;
  word-wrap: normal;
}

.c7 {
  box-sizing: border-box;
  padding-right: 15px;
  padding-left: 15px;
}

.c10 {
  box-sizing: border-box;
  padding-right: 15px;
  padding-left: 15px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 -15px;
}

.c9 {
  margin-bottom: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 14px;
  line-height: 1.1428571428571428;
}

.c8 {
  margin: auto;
  vertical-align: middle;
  padding-right: 4px;
}

.c0 {
  color: #0B0C0C;
  background-color: #FFFFFF;
  box-shadow: 0px 0px 4px 0px #505A5F;
  border-radius: 8px;
  display: none;
  z-index: 1;
  position: fixed;
  text-align: left;
  padding: 16px;
  white-space: normal;
  overflow: visible;
  max-width: 240px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1 {
  background-color: #FFFFFF;
  box-shadow: 0px 0px 4px 0px #505A5F;
  border-radius: 2px;
  display: block;
  z-index: 1;
  position: absolute;
  width: 16px;
  height: 16px;
  left: -8px;
  top: 50%;
  -webkit-transform: translateX(1px) translateY(-8px) rotate(45deg);
  -ms-transform: translateX(1px) translateY(-8px) rotate(45deg);
  transform: translateX(1px) translateY(-8px) rotate(45deg);
  border: none;
}

.c2 {
  background-color: #FFFFFF;
  display: block;
  z-index: 2;
  position: absolute;
  width: 16px;
  height: 24px;
  left: 0px;
  top: 50%;
  -webkit-transform: translateY(-12px);
  -ms-transform: translateY(-12px);
  transform: translateY(-12px);
}

.c4 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 14px;
  line-height: 1.1428571428571428;
}

.c5 {
  margin-bottom: 0;
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 20px;
  }
}

@media only screen and (min-width:641px) {
  .c7 {
    width: 12px;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
  }
}

@media only screen and (min-width:641px) {
  .c10 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-shrink: 1;
    -ms-flex-negative: 1;
    flex-shrink: 1;
  }
}

@media only screen and (min-width:641px) {
  .c6 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media print {
  .c9 {
    font-size: 14px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c9 {
    font-size: 16px;
    line-height: 1.25;
  }
}

@media print {
  .c4 {
    font-size: 14px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    font-size: 16px;
    line-height: 1.25;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    />
    <div
      class="c2"
    />
    <div
      class="c3 c4 c5"
    >
      <p>
        <strong>
          Sesame Street
        </strong>
      </p>
    </div>
    <div
      class="c3 c4 c5"
    >
      <p>
        1966
      </p>
    </div>
    <div
      class="c3 c4"
    >
      <p>
        Cookies eaten by the cookie monster compared to total US production
      </p>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7 c8"
      >
        <div
          class="c3 c9"
        >
          <p>
            âœ•
          </p>
        </div>
      </div>
      <div
        class="c10"
      >
        <div
          class="c3 c9"
        >
          <p>
            No data available
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
